### DISCUSSION TITLE
**Next.js error page (404, 400, 500) reload infinitely in development**

### DISCUSSION THREAD

**Discussion by Senbonzakura1234** [vercel/next.js#40000] [Help] [answered]
### Provide environment information
 - System:
    OS: Windows 10 10.0.22622
    CPU: (12) x64 Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
    Memory: 5.87 GB / 15.87 GB
 - Binaries:
    Node: 16.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 3.2.2 - C:\Program Files\nodejs\yarn.CMD
    npm: 8.16.0 - C:\Program Files\nodejs\npm.CMD

### Describe the Issue

I'm not sure if this is a Next.js bug or a function, but the app keep reloading when I'm on 404 or any other error page and it's really annoying and it start laggy when the log was long enough, I want to investigate a way to stop this.

### To reproduce

- Create a Next app
- Navigate to a page that doesn't exist to get a 404 error
- Open F12
- `on-demand-entries-client.js` script keep reloading the page

### Screen Record
[File](https://user-images.githubusercontent.com/43206347/187018884-3a16f7ea-5eb0-4b3d-9fb1-4dcd4d90b07b.mp4)

**Comment by icyJoseph** [comment, marked as answer]
I tried to control this with `onDemandEntries` on the next configuration file, but it didn't work, so I went to the source code, and this bit:

```js
            // don't attempt fetching the page if we're already showing
            // the dev overlay as this can cause the error to be triggered
            // repeatedly
            if (payload.event === 'pong' && payload.invalid && !self.__NEXT_DATA__.err) {
                // Payload can be invalid even if the page does exist.
                // So, we check if it can be created.
                fetch(location.href, {
                    credentials: 'same-origin'
                }).then((pageRes)=>{
                    if (pageRes.status === 200) {
                        // Page exists now, reload
                        location.reload();
                    } else {
                        // Page doesn't exist
                        if (self.__NEXT_DATA__.page === _router.default.pathname && _router.default.pathname !== '/_error') {
                            // We are still on the page,
                            // reload to show 404 error page
                            location.reload();
                        }
                    }
                });
            }
```

`// Page exists now, reload`, lead me to believe that this is done on purpose, because, during development, if you land on a page that doesn't exist in the file system, perhaps you are about to create it, or will create it, so it tries to pick up on that on intervals.

This is the nature of `_error` pages built like this.

However, if we for example return `{ notFound: true }` in GetServerSideProps, then Next.js understands that this is expected, and doesn't attempt to reload over and over.

Still, would be nice to be able to at least throttle this behavior.

**Comment by studiolama** [comment]
I am also facing same error. I have created a custom page for not-found but due to this is reloading again and again . Any solution ?

**Reply by mrpitch** [reply]
Same for me, the "normal" 404 page is also infinitely reloading. I'm using next.js 13.4.4

**Reply by michaelhpet** [reply]
I'm also experiencing the same thing on 13.4.4.

I think @icyJoseph's comment explains it. I thought it was a bug at first or something wrong with my implementation but it wasn't. It only happens in dev and not in production.

I think this should be addressed in the [docs](https://nextjs.org/docs/app/api-reference/file-conventions/not-found) @timneutkens

**Comment by bennik88** [comment]
Same error with "next": "13.4.4".

**Comment by Ian-Munashe** [comment]
Same error with "next": "13.4.5"

**Comment by abcdefghijorngarbosaxyz** [comment]
Same bug in not-found with "next": "13.4.6"

**Reply by NK3678** [reply]
please let me know, if you get answer. as for me also. not-found page is keep on loading on wrong route

**Comment by VinnyLima** [comment]
Same bug in not-found with "next": "13.4.6"

**Comment by tj-noor** [comment]
I ended up adding a catch-all route where the 404 was being generated from to return `notFound()` from it.

1. Create a folder: e.g. `[...404]`
2. Create a `page.tsx` or `page.jsx` page with the following:
```ts
import { notFound } from "next/navigation"

export default function NotFound() {
  return notFound()
}
```

**Reply by ThiagoPedais** [reply]
thank you

**Comment by Ian-Munashe** [comment]
I found a way which worked for me on **"next": 13.4.6**

- In your app directory just create a folder named `[...not_found]`.
- Inside that new folder create a file `page.tsx` or `page.jsx` and add this code:

```js
import { notFound } from "next/navigation";

const NotFoundCatchAll = () => notFound();

export default NotFoundCatchAll;

```

- In the root of your app directory create a file named `not-found.tsx` or `not-found.jsx` and add this code:

```js
export default function NotFound({}: Props) {
  return (
    <div>NotFound</div>
  );
}
```

Your custom 404 error page will work fine without the infinity reload issue.

**Reply by cheeselemon** [reply]
works like a charm ! 👍

**Reply by AndiZandi** [reply]
TY!

**Reply by VinnyLima** [reply]
Thanks, it work!!

**Reply by abcdefghijorngarbosaxyz** [reply]
> Your custom 404 error page will work fine without the infinity reload issue.

@NK3678

**Reply by inyee786** [reply]
This work,
@Ian-Munashe, can you please explain here, what is actually happening in the background if you have any idea of it?

**Reply by fuaditrockz** [reply]
Thanks it works!

**Reply by jmilbach** [reply]
Thank you, this solution worked well and it is versatile. I was able to add new pages/routes with no conflicts. Great little work around until the team gets a chance to address the issue.

**Reply by Jordaneisenburger** [reply]
For me this didn't work on 13.4.12

**Reply by chebread** [reply]
@Jordaneisenburger, Me too, this didn't work for me on 13.4.19.

**Reply by petrosmm** [reply]
does anyone have any idea why this route gets invoked while loading dynamic pages in `app` routes? i can't always tell if theres an actual error

**Reply by p00000001** [reply]
@petrosmm Apparently the 404 page is intentionally rendered "in anticipation". I raised a bug that our 404 page is always rendered without a 404 error, but it was closed as expected behaviour: #53252

**Reply by petrosmm** [reply]
@p00000001 does this mean i have to mark any potential errors with a guid or some other unique identifier? because i cant seem to throw or invoke notfound().... and redirect to it?

**Reply by zineanteoh** [reply]
works like magic ✨ (using next 14.0.3) 

thank you!

**Comment by manuelbasanta** [comment]
I tried to "debug" this "error" for an hour, how is this not commented [here](https://nextjs.org/docs/app/api-reference/file-conventions/not-found) in the docs ?

**Comment by manuelbasanta** [comment]
Fixed here: https://github.com/vercel/next.js/pull/51637

Working fine for me in V 13.4.12

**Reply by kdb13** [reply]
Not working for me in `v13.4.7`. Using the workaround by @Ian-Munashe.

**Comment by Johnjx** [comment]
The @Ian-Munashe fix works for me.
- "next": "13.4.7",

**Comment by BenjaminRiderelli** [comment]
I think its a dev bug only, I was having the same trouble, but when I pushed it (dont know how I thought about it) It worked perfectly on production

**Comment by ztanner** [comment]
Hey folks -- we've removed the dev server polling behavior on the v13.4.20 canary (specifically as of [v13.4.20-canary.4](https://github.com/vercel/next.js/releases/tag/v13.4.20-canary.4)). You should no longer be seeing infinite reloads, but if you do, please consider filing an issue for it with a reproduction so we can investigate.

**Reply by kdb13** [reply]
Could you please mention the stable version after which this is fixed, as I am running the version `13.4.7` and still facing the problem?

**Reply by ztanner** [reply]
13.4.20-canary corresponds with the 13.4.20 stable release.

**Comment by Jordaneisenburger** [comment]
This still happens on 13.5.6 even without a custom `not-found.tsx` file in `/src/app`, the native nextjs 404 is still infinitely reloading. Any ideas or others experiences this without the fixes supplied above ?

**Reply by timneutkens** [reply]
Please open a new issue with a reproduction 👍 Thank you!

**Reply by JurajKavka** [reply]
I tested Next 14 and seems to be OK (no repetitive requests). 3.4.12 was definitely broken.

**Reply by Jordaneisenburger** [reply]
Seems to be related with using generateStaticParams at the top of your page.
```
export const generateStaticParams = () => {
    return [];
};
```
and manually triggering `notfound()`, will create a reproduction repo soon.

**Comment by Gaurav-r-aj** [comment]
Update the next.js version in your project , it will fix the issue insted using next.js 13 use next.js 14

**Comment by OmarAfet** [comment]
Same error here with nextjs version 14.1

**Reply by gayancharith** [reply]
same error for 14.1.  Not only for the Not found page, but for every page of the applicaton.

**Reply by OmarAfet** [reply]
it looks like deleting the .next folder solves the problem for me.

but it still can happen sometimes

**Reply by ketijakrista** [reply]
Downgrading to 14.0.4 fixed for me

**Reply by OmarAfet** [reply]
In the latest version 13, I've been experiencing this issue. I'm not sure if it's related, but the problem began when I switched to macOS.

**Reply by ragnarok22** [reply]
I confirm. I'm facing the same issue with NextJS 14.1.0

**Comment by petrosmm** [comment]
can confirm this issue was fixed in 14.1.4 and broken again in 14.2.3. Still haven't found a combo of fixes to get this madness to stop. Seeing something similar to this https://github.com/vercel/next.js/issues/62228. though not sure if related?

**Comment by akbulatesra** [comment]
Next project with page router, it has 404.js but it is refreshing infinity. I saw this issue (https://github.com/vercel/next.js/issues/51132) but cant fix.

This issue has already been triaged, we're going to work on it, there's other issues that have a higher priority though. Please don't post further comments if there is no additional information being provided. That includes not posting "version x still has this", it's not helpful, the issue is still open so you can assume it's not fixed in version x. Every comment pings everyone on the thread and takes time away from us investigating/fixing this issue. These will be hidden automatically.

Originally posted by @timneutkens in https://github.com/vercel/next.js/issues/10024#issuecomment-1639747245

Is it same error or different? If it is same did you fix it?  @timneutkens

**Reply by timneutkens** [reply]
Looks like the same issue yes.

**Comment by Mangor1no** [comment]
This issue is still persist in latest NextJS version (15.3.0). I'm using page router, and unfortunately I cannot provide the source code atm.

**Comment by ErickTorresBrownU** [comment]
Another dev joining the mass grave here. 🙋‍♂️

**Reply by ragnarok22** [reply]
Welcome. Please grab a chair 🪑 and a coffee ☕.

**Reply by HafidzNurmawanHakim** [reply]
Another another dev joining the mass grave here. 🙋‍♂️

**Reply by GeorgeNabilPro** [reply]
That's really annoying, and I have no idea why such an issue has no clear solution until now.

**Comment by Ian-Munashe** [comment]
This issue was fixed long back. Just create a not-found tsx or jsx file in your app dir if you're using app router and create a starndard nextjs page function. Make sure you're importing asset starting with / through out the whole project. I you don't import your assets starting with /, your app will enter in a loop when it tries to load an asset that is not present
